### PR TITLE
Add ability to write extracted data back to ULog file

### PR DESCRIPTION
**Pull request by ken-voly**
Adds a function, `write_ulog()` which can write the currently extracted ulog data back to a ulog file. One potential use case of this feature is modifying a ulog file before replaying it in PX4. I've also added a unit test to check that the result of extracting, writing, and extracting again is equivalent to the original extracted data.

Currently, initial parameters and tagged logged messages are not supported because I don't have any data to test these features with.

Here's an example of the feature working. I used it to add some fake noise to a topic:

sensor_combined_z_accel_original
sensor_combined_z_accel_modified

### CODE REVIEW

**Review by JacobCrabill — APPROVED** [review]
Looks good!  @bkueng mind taking a look?

File: pyulog/core.py [hunk 1]
```diff
@@ -234,12 +237,238 @@ def get_dataset(self, name, multi_instance=0):
         return [elem for elem in self._data_list
                 if elem.name == name and elem.multi_id == multi_instance][0]
 
+    def write_ulog(self, path):
+        """ write current data back into a ulog file """
+        ulog_file = open(path, "wb")
```
Comment (bkueng) [inline pyulog/core.py:242]:
```suggestion
        with open(path, "wb") as ulog_file:
```

File: test/test_ulog_write.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        # Check tagged logged string message 'C'
+        assert original.logged_messages_tagged == copied.logged_messages_tagged
+
+        # Check dropouts 'O'
+        assert original.dropouts == copied.dropouts
+
+        # Check changed parameters 'P'
+        assert original.changed_parameters == copied.changed_parameters
+
+        pass
```
Comment (bkueng) [inline test/test_ulog_write.py:67]:
```suggestion
```

**Review by bkueng — COMMENTED** [review]
Very cool, nice work!
I will have a closer look, and provide a sample log with the missing data, so we can add the remaining messages.

**Comment by bkueng** [comment]
I added a sample log with logging_tagged messages and default params: https://github.com/PX4/pyulog/tree/add_sample_log

File: pyulog/core.py [hunk 3]
```diff
@@ -252,6 +481,23 @@ def __init__(self, message_add_logged_obj):
             for name in np_array.dtype.names:
                 self.data[name] = np_array[name]
 
+        def __eq__(self, other):
+            if not isinstance(other, ULog.Data):
+                return NotImplemented
+
+            # Compare data arrays
+            data_equal = True
+            arrays_equal = data_equal and self.data.keys() == other.data.keys()
```
[the remark this answers is not in the document]
Comment (ken-voly) [reply pyulog/core.py]: Fixed

File: pyulog/core.py (hunk 1 above)
Comment (ken-voly) [reply pyulog/core.py]: Fixed

File: test/test_ulog_write.py (hunk 2 above)
Comment (ken-voly) [reply test/test_ulog_write.py]: Fixed

**Review by bkueng — COMMENTED** [review]
I fixed CI in https://github.com/PX4/pyulog/pull/68, if you rebase on master, it should pass as well.

File: pyulog/core.py [hunk 4]
```diff
@@ -234,12 +237,264 @@ def get_dataset(self, name, multi_instance=0):
...
+
+            file.write(header)
+            file.write(data)
+
+    def _write_default_parameters(self, file):
+        default_types = {}
+
+        for bit in self._default_parameters.keys():
+            for name, value in self._default_parameters[bit].items():
+                if name in default_types:
```
[the remark this answers is not in the document]
Comment (ken-voly) [reply pyulog/core.py:324]: I see. I think it will be easiest to write them back separately.

**Comment by bkueng** [comment]
I cleaned the commits, increased the pylint limits a bit (we'll need to split it at some point) and merged it.

**Pull request closed without merging**